### ISSUE TITLE
live example (actix-web) cant receive activities from mastodon with Error(Incoming activity has invalid signature)

### ISSUE THREAD

**Issue by CutestNekoAqua** [LemmyNet/activitypub-federation-rust#107]
Hi, I played a bit with the example code and replaced axum with actix-web in the live example. The user can be fetched but no matter what fediverse software I try federating with (mastodon, glitch, firefish, akkoma) it all fails at verification of the incoming activity. The culprit code is [here](https://github.com/LemmyNet/activitypub-federation-rust/blob/main/src/http_signatures.rs#L186-L222)

**Comment by CutestNekoAqua** [comment]
There might be a bug in my code, but I am not 100%-ly sure and being at this stuck for at least two hours now makes me wonder if this is related to Mastodon having a bug in its signing code. And this check being bug-free ;')

**Comment by Nutomic** [comment]
Lemmy uses the same code and can federate with Mastodon just fine. So I doubt that there is a bug in the library or in Mastodon. Is it possible that its verifying with the wrong public key for some reason?

**Comment by CutestNekoAqua** [comment]
> Lemmy uses the same code and can federate with Mastodon just fine. So I doubt that there is a bug in the library or in Mastodon. Is it possible that its verifying with the wrong public key for some reason?

I mean, im not even sure about how the lib even gets any public key. Afaik thats not clearly documented in the example. I guess you mean, that the local instance is started with the wrong public key?

**Comment by CutestNekoAqua** [comment]
> Another problem I had recently with receiving activities was that the nginx config didnt set the Host header. Its important that this matches the federation domain as its part of the signature

that was the call I needed! Thanks!

**Comment by Technohacker** [comment]
I stumbled upon this error in a subtly different way, and I figured I'd add to this issue for archival purposes:

I'm using `axum` and my inbox endpoint is at `/user/<user id>/inbox`, but the message to be validated was as follows:
```
key
 -----BEGIN PUBLIC KEY-----
 (correct key)
 -----END PUBLIC KEY-----

message
 host: (correct hostname)
 date: (correct time)
 digest: (valid SHA256 signature)
 content-type: application/activity+json
 (request-target): post /abc/inbox
```

It took me a while before I noticed that the problem was the `(request-target)`. The `/user/` prefix was missing. I'd known axum's nested `Router`s would do such a thing, but I didn't realize the extractor within `ActivityData` was using the stripped URL.

Un-nesting the Router (for now), fixes the signature check. Maybe `ActivityData` could use `OriginalUri` instead, but I feel that might introduce SemVer-breaking behaviour.
